### PR TITLE
Add WIP Q3 Importer

### PR DESCRIPTION
**Pull request by krazyjakee**
Godot_v4 3-stable_win64_CCep14QxjQ

Picture shows "Oasis" from Wolfenstein: Enemy Territory with horribly broken mesh and missing textures (random colors assigned).

The franken-bsp-reader file needs a big refactor so I'll just fudge all the BSP3 stuff onto the end and fix what I can.
Was there a reason you want all this stuff in one file?

- [x] ~~BSP3 detection~~
- [x] ~~Lump parsing~~
- [ ] Reuse existing logic (Existing classes and methods)
- [ ] Fix mesh generation
- [ ] Handle large meshes (Some maps are too large for SurfaceTool)
- [ ] Apply/Create textures
- [ ] Create colliders
- [ ] Create entities

### CODE REVIEW

**Comment by krazyjakee** [comment]
@CSLRDoesntGameDev Any interest in helping out with this one?

**Comment by CSLRDoesntGameDev** [comment]
yeah i wouldn't mind helping, however with how cluttered the script is it might be better to for a whole script refactor - up to jitspoe though.

between the original mess then the added mess from my quake 2 importer it's probably gonna be a lot to merge the code together

**Comment by krazyjakee** [comment]
@CSLRDoesntGameDev thanks, I'll see how far I can get with a refactor

**Comment by CSLRDoesntGameDev** [comment]
yeah @krazyjakee its gonna be fun to do!

I know jitspoe said he wanted to have it all be a single unified script which is probably fine for just a couple BSP Formats, which is why my quake 2 importer was a separate script, to avoid the clutter of multiple formats.

I Wasn't super happy with the original q2 script but it worked so i just pushed it, Quake 2 is such a complex format that a lot of it would just be for quake 2.

Quake 2 and Quake 3 are nearly identical formats, you could 100% leverage the Quake 2 code for Quake 3 with the same code structure since my importer compiled the vertices directly in each face for easier mesh construction, Quake 3 does the same thing so that part could definitely be the same.

File: addons/bsp_importer/bsp_reader.gd [hunk 1]
```diff
@@ -1,64 +1,89 @@
...
 
 class BSPEdge:
 	var vertex_index_0 : int
 	var vertex_index_1 : int
+
 	func read_edge_16_bit(file : FileAccess) -> int:
 		vertex_index_0 = file.get_16()
 		vertex_index_1 = file.get_16()
-		return 4 # 2 2 byte values
+		return 4
```
Comment (jitspoe) [inline addons/bsp_importer/bsp_reader.gd:81]: Why are you removing the comments explaining what these are?

**Comment by jitspoe** [comment]
The reason I wanted it to all be one unified thing is for maintainability when adding/tweaking features as well as just avoiding a ton of duplicated code.  The Quake bsp formats are like 90+% the same with a few differences in each version, such as additional lumps or different limits.  Having a unique importer for each BSP version would be unwieldy.  Each feature (generating materials, mapping entities, importing lights, adding occlusion culling, etc.) would all have to be independently implemented.  To me, it makes more sense to just import all the formats through one path and have slight differences based on the format (ex: use 32 bit values in some cases and 16bit in others).

Also, regarding this change in general, it seems like there are a ton of changes that aren't relevant to the Q3 bsp importer, like removing comments and changing variable to be untyped as well as general formatting changes that make the diff very difficult to parse.  Also, this isn't going to merge well with the changes to bring in Q2 bsp support.  I think it might be better to re-start with that as a base and just bring in the Q3 specific changes and don't change the other code unless it needs an actual logic change so the merge will be cleaner.

Thanks for giving this a go!

**Comment by krazyjakee** [comment]
I agree :+1:
thanks for taking the time to review

**Pull request closed without merging**